File: Project-in-Java/src/odev/Oyun.java
package odev;

import java.util.Iterator;
import java.util.Random;
import java.util.Scanner;

public class Oyun {
	
	Scanner scanner = new Scanner(System.in);
		
	
	public void basla() { 
		
		
		
		
		
		String str = scanner.nextLine();
        String[] strArray = str.split(" ");
        int[] intArray = new int[strArray.length];

        for (int i = 0; i < strArray.length; i++) {
            intArray[i] = Integer.parseInt(strArray[i]);
        }
        
        Koloni [] k = new Koloni[intArray.length];
        
        char [] s = new char [intArray.length] ;
        
        for(int i =65 , j =0; i< 65+intArray.length;i++,j++) {
        	s[j]=(char)i;
        }

        for (int i = 0; i < intArray.length; i++) {
        	
        	k[i]=new Koloni(intArray[i],s[i]);
            //System.out.print(intArray[i] + " \n");
        }
        
        //simge çevirme
       /*int a =65;
       char b = (char)a;
       
       System.out.println(b);*/
        
        
       
       //koloni yazdýrma
       // for(int i=0;i<intArray.length;i++ ) {
    	//   System.out.println(k[i]);
       //}
       
       
       
       /*
       //4 populasyon varsa
       for(int i =0 ; i<3 ;i++) { //  i =lengt-1 sýnýr
    	   for(int j=i+1 ; j<4;j++) { // j=lengt sýnýr
    		   
    		   System.out.println(i + " "+j);
    	   }
       }*/
        
        
        ATaktik a = new ATaktik();
        BTaktik b = new BTaktik();
        
        AUretim a1 = new AUretim();
        BUretim b1 = new BUretim();

        
        int [] puan = new int[15];
        
        int [] uret = new int[15];
       
        
        Random random = new Random();
        //int r = random.nextInt(2);
        
        //System.out.println(puan[r]);
       
       
       int tur =0;
       
       while(true) {
    	   
    	   System.out.println("-------------------------------------------------------");
           System.out.println("Tur Sayisi:"+tur);
           System.out.println("Koloni   Populasyon   Yemek Stogu   Kazanma  Kaybetme");
           for(int i=0;i<intArray.length;i++ ) {
           	System.out.println(k[i]);
           }
           System.out.println("\n-------------------------------------------------------");
    	   
    	   
    	   for(int i =0 ; i<intArray.length-1 ; i++) { //  i =lengt-1 sýnýr
        	   for(int j=i+1 ; j<intArray.length ; j++) { // j=lengt sýnýr
        		   
        		   if(k[i].getPopulasyon() == 0 || k[j].getPopulasyon() == 0) {// elenen koloni olunca savasý olmayacak
        			   continue;
        		   }
        		   
        		   
        		   
        		   for(int q =0,z=1;q<15 && z<15  ;q+=2,z+=2) { 
        			   puan[q]=a.Savas();   
        			   puan[z]=b.Savas(); 
        		   }
        		   
        		   int r = random.nextInt(15);
        		   int t = random.nextInt(15);
        		   
        		   k[i].setPuan(puan[r]);
        		   k[j].setPuan(puan[t]);
        		   
        		   int kazanmaI = k[i].getKazanma()+1;
        		   int kazanmaJ = k[j].getKazanma()+1;
        		   int kaybetmeI = k[i].getKaybetme()+1;
        		   int kaybetmeJ = k[j].getKaybetme()+1;
        		   
        		   
        		   
        		   
        		   System.out.println(k[i].getPuan()+","+k[j].getPuan());

        		   
        		   if(k[i].getPuan()>k[j].getPuan()) {//k[i] kazanýr
        			   
        			   int fark = k[i].getPuan() - k[j].getPuan();
        			   int oran = fark*100/1000;
        			   k[j].setPopulasyon(k[j].getPopulasyon() - (k[j].getPopulasyon()*oran/100));
        			   k[j].setYemekStogu(k[j].getYemekStogu() - (k[j].getPopulasyon()*oran/100));
        			   k[i].setYemekStogu(k[i].getYemekStogu() + (k[j].getPopulasyon()*oran/100));
        			   
        			  
        			   
        			   k[i].setKazanma(kazanmaI);
        			   k[j].setKaybetme(kaybetmeJ);
        			   
        		   }
        		   else if(k[i].getPuan()<k[j].getPuan()) {//k[i] kaybeder
        			   
        			   
        			   int fark = k[j].getPuan() - k[i].getPuan();
        			   int oran = 1000/fark;
        			   k[i].setPopulasyon(k[i].getPopulasyon() - (k[i].getPopulasyon()*oran/100));
        			   k[i].setYemekStogu(k[i].getYemekStogu() - (k[i].getPopulasyon()*oran/100));
        			   k[j].setYemekStogu(k[j].getYemekStogu() + (k[i].getPopulasyon()*oran/100));
        			   
        			   
        			   
        			   k[j].setKazanma(kazanmaJ);
        			   k[i].setKaybetme(kaybetmeI);
        			   
        		   }
        		   else {//berabere
        			   
        			   int berabere = random.nextInt(2);
        			   
        			   
        			   if(berabere == 0) {//k[i] kazansýn
        				   
        				   int fark = k[i].getPuan() - k[j].getPuan() + 100;
            			   int oran = fark*100/1000;
            			   k[j].setPopulasyon(k[j].getPopulasyon() - (k[j].getPopulasyon()*oran/100));
            			   k[j].setYemekStogu(k[j].getYemekStogu() - (k[j].getPopulasyon()*oran/100));
            			   k[i].setYemekStogu(k[i].getYemekStogu() + (k[j].getPopulasyon()*oran/100));
            			   
        				   
            			   
            			   k[i].setKazanma(kazanmaI);
            			   k[j].setKaybetme(kaybetmeJ);
        				   
        			   }
        			   else if(berabere == 1) {//k[j] kazansýn
        				   
        				   int fark = k[j].getPuan() - k[i].getPuan() +100;
            			   int oran = fark*100/1000;
            			   k[i].setPopulasyon(k[i].getPopulasyon() - (k[i].getPopulasyon()*oran/100));
            			   k[i].setYemekStogu(k[i].getYemekStogu() - (k[i].getPopulasyon()*oran/100));
            			   k[j].setYemekStogu(k[j].getYemekStogu() + (k[i].getPopulasyon()*oran/100));
        				   
            			   
            			   k[j].setKazanma(kazanmaJ);
            			   k[i].setKaybetme(kaybetmeI);
        				   
        			   }
        			   else {
        				   ;
        			   }
        			   
        		   }
        			   
        		   
        		   
        		   
        		   
        		   
        		   
        		   
        		   
        		   
        		   
        		   System.out.println(i + " "+j);
        	   }
        	
        	   
    	   }
    	   
    	   
    	   for(int q =0,z=0;q<15 && z<15;q+=2,z+=2) { 
			   uret[q]=a1.Uret();   
			   uret[z]=b1.Uret(); 
		   }
    	   
    	   
    	 //tur sonu olanlar
    	   for(int w =0;w<intArray.length;w++) {
    		   
    		   int uSecim = random.nextInt(15);
    		   int uOnceki= k[w].getYemekStogu();
    		   k[w].setYemekStogu(uOnceki+uret[uSecim]);// her tur 1-10 yemek ekle stoða
    		   
    		   if(k[w].getPopulasyon() <= 0 || k[w].getYemekStogu() <=0 ) {// koloni elenirse
    			   k[w].setPopulasyon(0);
    			   k[w].setYemekStogu(0);
    			   k[w].setKazanma(0);
    			   k[w].setKaybetme(0);
    		   }
    		   
    		   if(k[w].getPopulasyon() != 0 || k[w].getYemekStogu() != 0){// pop ve stok iþlemleri
    			 //pop
    			   int pop = k[w].getPopulasyon();
    			   k[w].setPopulasyon(pop + (pop*20/100));
    			 //stok  
        		   int stok = k[w].getYemekStogu() -pop*2;
        		   k[w].setYemekStogu(stok);
    		   }
    		   else {
    			   ;
    		   }
    		   
    		 
    	   }
        	   
        tur++;
        
        int sayi =0;
        for (int w =0;w<intArray.length;w++) {
        	if(k[w].getPopulasyon() == 0){
        		sayi++;
        	}
        	
        }
        
        if(sayi==(intArray.length-1)) break;  
        
       }
       
       
       
       
       System.out.println("-------------------------------------------------------");
       System.out.println("Tur Sayisi:"+tur);
       System.out.println("Koloni   Populasyon   Yemek Stogu   Kazanma  Kaybetme");
       for(int i=0;i<intArray.length;i++ ) {
       	System.out.println(k[i]);
       }
       System.out.println("\n-------------------------------------------------------");
       
       
       
       
       	
		
	}

}

File: Project-in-C/src/Oyun.c
#include "Oyun.h"
#include "Koloni.h"


Oyun OyunOlustur() {
	Oyun this;
	this = (Oyun)malloc(sizeof(struct OYUN));
	this->basla = &basla;
	return this;
}
void basla(const Oyun this) {

	char str[50];
	int koloniSayisi = 0;
	int pops[30];

	printf("Populasyonlari girin: ");
	fgets(str, sizeof(str), stdin);
	char* token = strtok(str, " ");

	while (token != NULL) {
		
		pops[koloniSayisi] = atoi(token);
		koloniSayisi++;
		token = strtok(NULL, " ");
	}

	Koloni k[koloniSayisi];

	for (int i = 0; i < koloniSayisi; i++) {
		k[i] = KoloniOlustur(pops[i], 65+i);
	}
	int tur = 0;

	Ataktik a = AtaktikOlustur();
	Btaktik b = BtaktikOlustur();

	Auretim a1 = AuretimOlustur();
	Buretim b1 = BuretimOlustur();


	int puan[30];

	int uret[14];
	

	while (1) {
		printf("\n");
		printf("------------------------------------------------------------------");
		printf("\n");

		printf("Tur Sayisi: %d\n ",tur);
		
		printf("% s\n", "Koloni   Populasyon   YemekStogu   Kazanma   Kaybetme");

		for (int i = 0; i < koloniSayisi; i++) {
			
			k[i]->toString(k[i]);
			printf("\n");

		}

		for (int i = 0; i < koloniSayisi - 1; i++) {
			for (int j = i + 1; j < koloniSayisi; j++) {
				if (k[i]->getPopulasyon(k[i]) == 0 || k[j]->getPopulasyon(k[i]) == 0) {
					continue;
				}

				for (int q = 0, z = 1; q < 31 && z < 30; q += 2, z += 2) {
					puan[q] = a->super->savas(a,q+1);//hep ayni sayi
					puan[z] = b->super->savas(b,z+2);
					
				}

				//for (int i = 0; i < 14; i++) {
				//	printf("\n%d",puan[i]);
				//}

				int random1;
				int random2;

				// Seed değerini zamanla güncelleme
				srand(time(0));

				// 0 ile 15 arasında rastgele bir değer üretme
				//hep ayni sayi

				random1 = (rand() % 4) + j + i;
				random2 = (rand() % 9) + j + 2;


				k[i]->setPuan(k[i],puan[random1]);
				k[j]->setPuan(k[j], puan[random2]);

				int kazanmaI = (k[i]->getKazanma(k[i])) + 1;
				int kazanmaJ = (k[j]->getKazanma(k[j])) + 1;
				int kaybetmeI = (k[i]->getKaybetme(k[i])) + 1;
				int kaybetmeJ = (k[j]->getKaybetme(k[j])) + 1;

				
				printf("\n%d %d", k[i]->getPuan(k[i]), k[j]->getPuan(k[j]));

				if (k[i]->getPuan(k[i]) > k[j]->getPuan(k[j])) {//k[i] kazanýr

					int fark = k[i]->getPuan(k[i]) - k[j]->getPuan(k[j]);
					int oran = fark * 100 / 1000;
					k[j]->setPopulasyon(k[j],(k[j]->getPopulasyon(k[j]) - (k[j]->getPopulasyon(k[j])) * oran / 100));
					k[j]->setYemekStogu(k[j], (k[j]->getYemekStogu(k[j]) - (k[j]->getPopulasyon(k[j])) * oran / 100));
					k[i]->setYemekStogu(k[i],(k[i]->getYemekStogu(k[i]) + (k[j]->getPopulasyon(k[j])) * oran / 100));



					k[i]->setKazanma(k[i],kazanmaI);
					k[j]->setKaybetme(k[j],kaybetmeJ);

				}
				else if (k[i]->getPuan(k[i]) < k[j]->getPuan(k[j])) {//k[i] kaybeder


					int fark = k[j]->getPuan(k[j]) - k[i]->getPuan(k[i]);
					int oran = fark * 100 / 1000;
					k[i]->setPopulasyon(k[i],(k[i]->getPopulasyon(k[i]) - (k[i]->getPopulasyon(k[i])) * oran / 100));
					k[i]->setYemekStogu(k[i], (k[i]->getYemekStogu(k[i]) - (k[i]->getPopulasyon(k[i])) * oran / 100));
					k[j]->setYemekStogu(k[j],(k[j]->getYemekStogu(k[j]) + (k[i]->getPopulasyon(k[i])) * oran / 100));



					k[j]->setKazanma(k[j],kazanmaJ);
					k[i]->setKaybetme(k[i],kaybetmeI);

				}
				else {//berabere

					int berabere = rand() % 2;


					if (berabere == 0) {//k[i] kazansýn

						int fark = k[i]->getPuan(k[i]) - k[j]->getPuan(k[j]) + 100;
						int oran = fark * 100 / 1000;
						k[j]->setPopulasyon(k[j], (k[j]->getPopulasyon(k[j]) - (k[j]->getPopulasyon(k[j])) * oran / 100));
						k[j]->setYemekStogu(k[j], (k[j]->getYemekStogu(k[j]) - (k[j]->getPopulasyon(k[j])) * oran / 100));
						k[i]->setYemekStogu(k[i], (k[i]->getYemekStogu(k[i]) + (k[j]->getPopulasyon(k[j])) * oran / 100));



						k[i]->setKazanma(k[i], kazanmaI);
						k[j]->setKaybetme(k[j], kaybetmeJ);

					}
					else if (berabere == 1) {//k[j] kazansýn

						int fark = k[j]->getPuan(k[j]) - k[i]->getPuan(k[i]) + 100;
						int oran = fark * 100 / 1000;
						k[i]->setPopulasyon(k[i], (k[i]->getPopulasyon(k[i]) - (k[i]->getPopulasyon(k[i])) * oran / 100));
						k[i]->setYemekStogu(k[i], (k[i]->getYemekStogu(k[i]) - (k[i]->getPopulasyon(k[i])) * oran / 100));
						k[j]->setYemekStogu(k[j], (k[j]->getYemekStogu(k[j]) + (k[i]->getPopulasyon(k[i])) * oran / 100));



						k[j]->setKazanma(k[j], kazanmaJ);
						k[i]->setKaybetme(k[i], kaybetmeI);
					}
					else {
						
					}

				}


			}

		}

		for (int q = 0, z = 0; q < 15 && z < 14; q += 2, z += 2) {
			uret[q] = a1->super->uret(a1);
			uret[z] = b1->super->uret(b1);
		}

		//tur sonu 
		for (int w = 0; w < koloniSayisi; w++) {

			int uSecim = rand() % 14;
			int uOnceki = k[w]->getYemekStogu(k[w]);
			k[w]->setYemekStogu(k[w], (uOnceki + uret[uSecim]));


			if (k[w]->getPopulasyon(k[w]) != 0 && k[w]->getYemekStogu(k[w]) != 0) {
				//pop
				int pop = k[w]->getPopulasyon(k[w]);
				k[w]->setPopulasyon(k[w], pop + (pop * 20 / 100));
				//stok  
				int stok = k[w]->getYemekStogu(k[w]) - pop * 2;
				k[w]->setYemekStogu(k[w], stok);
			}

			if (k[w]->getPopulasyon(k[w]) <= 0 || k[w]->getYemekStogu(k[w]) <= 0) {
				k[w]->setPopulasyon(k[w],0);
				k[w]->setYemekStogu(k[w],0);
				k[w]->setKazanma(k[w],0);
				k[w]->setKaybetme(k[w],0);
			}
		}
		printf("------------------------------------------------------------------");
		printf("\n");

		tur++;

		int sayi = 0;
		for (int w = 0; w < koloniSayisi; w++) {
			if (k[w]->getPopulasyon(k[w]) == 0) {
				sayi++;
			}

		}

		if (sayi == (koloniSayisi - 1)) break;
	}
	printf("\n");
	printf("------------------------------------------------------------------");
	printf("\n");

	printf("Tur Sayisi: %d\n ", tur);

	printf("% s\n", "Koloni   Populasyon   YemekStogu   Kazanma   Kaybetme");

	for (int i = 0; i < koloniSayisi; i++) {

		k[i]->toString(k[i]);
		printf("\n");

	}
	printf("------------------------------------------------------------------");
	printf("\n");



}
void OyunYoket(const Oyun this) {

	free(this);
}
